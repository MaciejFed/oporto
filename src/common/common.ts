import { DateTime } from 'luxon';
import { Zone } from 'luxon/src/zone';
import { DurationLike } from 'luxon/src/duration';
import { Exercise } from '../exercise/exercise';

export class DateTimeExtended {
  private readonly date: DateTime;

  private constructor(date: Date) {
    this.date = DateTime.fromJSDate(date);
  }

  public plus(duration: DurationLike) {
    const newDate = this.date.plus(duration);
    return DateTimeExtended.fromJSDate(newDate.toJSDate());
  }

  public toJSDate(): Date {
    return this.date.toJSDate();
  }

  get ordinal(): number {
    return this.date.year * 365 + this.date.ordinal;
  }

  get weekday(): number {
    return this.date.weekday;
  }

  get weekNumber(): number {
    this.date.toJSDate();
    return this.date.weekNumber;
  }

  static fromJSDate(date: Date, options?: { zone?: string | Zone }): DateTimeExtended {
    return new DateTimeExtended(date);
  }

  static now(): DateTimeExtended {
    return new DateTimeExtended(new Date());
  }
}

export interface Comparable {
  equal(other: Comparable): boolean;
}

export function onlyDistinct(arr: Comparable[]): Comparable[] {
  const distinctElements: Comparable[] = [];
  for (let i = 0; i < arr.length; i++) {
    let unique = true;
    for (let j = 0; j < distinctElements.length; j++) {
      if (i !== j && arr[i].equal(distinctElements[j])) {
        unique = false;
      }
    }
    if (unique) {
      distinctElements.push(arr[i]);
    }
  }
  return distinctElements;
}

export function enforceArrayLimit(arr: any[], limit: number) {
  if (arr.length > limit) {
    arr.splice(limit);
  }
  return arr;
}

export function onlyDistinctSubjects(arr: Exercise[]): Exercise[] {
  const distinctElements: Exercise[] = [];
  for (let i = 0; i < arr.length; i++) {
    let unique = true;
    for (let j = 0; j < distinctElements.length; j++) {
      if (i !== j && arr[i].getBaseWord() === distinctElements[j].getBaseWord()) {
        unique = false;
      }
    }
    if (unique) {
      distinctElements.push(arr[i]);
    }
  }
  return distinctElements;
}

export function findWordStartIndex(sentence: string, word: string) {
  const words = sentence.split(' ');
  const index = words.indexOf(word);

  return words.slice(0, index).reduce((prev, curr) => curr.length + prev + 1, 0);
}

export function getRandomElement<T>(arr: T[], maxLength?: number): T {
  // eslint-disable-next-line no-nested-ternary
  const length = maxLength ? (arr.length > maxLength ? maxLength : arr.length) : arr.length;
  return arr[Math.floor(Math.random() * length)];
}

export async function sleep(milliseconds: number) {
  await new Promise((resolve) => setTimeout(resolve, milliseconds));
}

export function formatDate(date: Date): string {
  const dateTime = DateTime.fromJSDate(date);
  return `${date.getFullYear()}-${
    date.getMonth() + 1
  }-${date.getDate()} ${date.getHours()}:${date.getMinutes()} [${dateTime.toRelative()}]`;
}

export function isBeforeWeekday(resultDate: Date, weekday: number) {
  const now = DateTime.now();
  const dayInYearNow = (now.weekNumber - 1) * 7 + weekday;

  const resultDateTime = DateTime.fromJSDate(resultDate);
  const resultDayInYear = (resultDateTime.weekNumber - 1) * 7 + resultDateTime.weekday;

  return resultDayInYear <= dayInYearNow;
}

export function isOnWeekDay(resultDate: Date, weekday: number, year: number) {
  const now = DateTime.now();
  const resultDateTime = DateTime.fromJSDate(resultDate);

  return (
    now.weekNumber === resultDateTime.weekNumber && resultDateTime.weekday === weekday && resultDateTime.year === year
  );
}

export function areSameCalendarDay(date1: Date, date2: Date): boolean {
  return (
    date1.getFullYear() === date2.getFullYear() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getDate() === date2.getDate()
  );
}

export function assertNonNull<T>(value?: T): T {
  if (!value) throw new Error('Unexpected Undefined!');

  return value;
}
