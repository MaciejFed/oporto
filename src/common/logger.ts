import winston from 'winston';

const winstonLogger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
    winston.format.prettyPrint()
  ),
  transports: [new winston.transports.File({ filename: 'out.log' })]
});

export const logger = {
  info: (msg: string, ...arg: any[]) => {
    console.log(msg, arg);
    winstonLogger.info(msg);
  },
  debug: (msg: string, ...arg: any[]) => {
    console.debug(msg, arg);
    winstonLogger.debug(msg);
  },
  error: (msg: string, ...arg: any[]) => {
    console.error(msg, arg);
    winstonLogger.error(msg, arg);
  }
}
