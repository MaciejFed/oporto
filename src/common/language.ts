import { getRandomElement } from './common';

export enum Language {
  Portuguese = 'Portuguese',
  German = 'German'
}

export function getVoice(language: Language): string {
  if (language === Language.Portuguese) {
    return getRandomElement(['Joana', 'Joaquim']);
  }
  return getRandomElement(['Anna', 'Petra', 'Markus']);
}
