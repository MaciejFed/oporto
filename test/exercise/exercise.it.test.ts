/* eslint-disable @typescript-eslint/no-var-requires */
import { sleep } from '../../src/common/common';
import { simulateTyping } from '../util';

const output: string[] = [];
 // @ts-ignore
 process.stdin.setRawMode = () => {};

jest.mock('child_process', () => {
  return {
    __esModule: true,
    exec: (command: string) => {}
  };
});
jest.mock('terminal-kit', () => {
    return {
      __esModule: true,
      terminal: {
        moveTo: (x: number, y: number, data: string) => {
          if (data)
            output.push(data);
        },
        hideCursor: (hide: boolean) => {},
        green: () => {},
        red: () => {},
        white: () => {}
      }
    };
  });



type AppModules = {
  SessionManager: any;
  Input: any;
  Terminakl: any;
  eventProcessor: any;
};


function requireAllModules(): AppModules {
  const eventProcessor = require('../../src/event/eventProcessor').eventProcessor;
  const SessionManager = require('../../src/session/sessionManager').SessionManager;
  const Input = require('../../src/io/input').Input;
  const Terminakl = require('../../src/io/terminal').Terminakl;

  return {
    SessionManager: SessionManager,
    Input: Input,
    Terminakl: Terminakl,
    eventProcessor: eventProcessor
  };
}

describe('Exercises IT Snapshot', () => {
  beforeEach(() => {
    // @ts-ignore
    const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => {});
    output.length = 0;
  });

  afterEach(() => {
    jest.resetModules();
    process.stdin.removeAllListeners();
  });

  it('Translation', async () => {
    jest.mock('../../src/exercise/exercise', () => {
      const readAll = require('../../src/repository/exercisesRepository').readAll;
      const TranslationExercise = require('../../src/exercise/translationExercise').TranslationExercise;
      const modelActual = jest.requireActual('../../src/exercise/exercise');
      const translationExerciseToPortuguese = new TranslationExercise();
      translationExerciseToPortuguese.translation = readAll().translations[1];
      translationExerciseToPortuguese.translationType = 'toPortuguese';
      const translationExerciseToEnglish= new TranslationExercise();
      translationExerciseToEnglish.translation = readAll().translations[1];
      translationExerciseToEnglish.translationType = 'toEnglish';
      return {
        __esModule: true,
        modelActual,
        generateUniqeExercises: () => [translationExerciseToEnglish, translationExerciseToPortuguese]
      };
    });
    const appModules = requireAllModules();
    const myTerminal = new appModules.Terminakl(appModules.eventProcessor);
    const input = new appModules.Input(appModules.eventProcessor);
    const sessionManager = new appModules.SessionManager(
      appModules.eventProcessor,
      2
    );
    appModules.eventProcessor.emit('APP_STARTED');
    simulateTyping('o chÃ¡');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    simulateTyping('tea');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    const eventHistory = appModules.eventProcessor.eventHistory.map(
      (event: { event: any }) => event.event
    );
    expect(eventHistory).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });


  it('Fit In', async () => {
    jest.mock('../../src/exercise/exercise', () => {
      const readAll = require('../../src/repository/exercisesRepository').readAll;
      const FitInGapExercise = require('../../src/exercise/fitInGapExercise').FitInGapExercise;
      const modelActual = jest.requireActual('../../src/exercise/exercise');
      const fitInGapExerciseOne = new FitInGapExercise();
      const fitInGapExerciseTwo = new FitInGapExercise();
      fitInGapExerciseOne.fitIn = readAll().fitIn[0];
      fitInGapExerciseTwo.fitIn = readAll().fitIn[1];
      return {
        __esModule: true,
        modelActual,
        generateUniqeExercises: () => [fitInGapExerciseOne, fitInGapExerciseTwo]
      };
    });
    const appModules = requireAllModules();
    const myTerminal = new appModules.Terminakl(appModules.eventProcessor);
    const input = new appModules.Input(appModules.eventProcessor);
    const sessionManager = new appModules.SessionManager(
      appModules.eventProcessor,
      2
    );
    appModules.eventProcessor.emit('APP_STARTED');
    simulateTyping('de');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    simulateTyping('wronganswer');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    const eventHistory = appModules.eventProcessor.eventHistory.map(
      (event: { event: any }) => event.event
    );
    expect(eventHistory).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });

  it('Verb', async () => {
    jest.mock('../../src/exercise/exercise', () => {
      const readAll = require('../../src/repository/exercisesRepository').readAll;
      const RegularVerbExercise = require('../../src/exercise/verbExercise').RegularVerbExercise;
      const IrregularVerbExercise = require('../../src/exercise/verbExercise').IrregularVerbExercise;
      const modelActual = jest.requireActual('../../src/exercise/exercise');
      const regularVerbExercise = new RegularVerbExercise();
      const irregularVerbExercise = new IrregularVerbExercise();
      regularVerbExercise.verb = readAll().verbs.regular[0];
      regularVerbExercise.person = 'Eu'
      irregularVerbExercise.verb = readAll().verbs.irregular[0];
      irregularVerbExercise.person = 'Eu'
      return {
        __esModule: true,
        modelActual,
        generateUniqeExercises: () => [regularVerbExercise, irregularVerbExercise]
      };
    });
    const appModules = requireAllModules();
    const myTerminal = new appModules.Terminakl(appModules.eventProcessor);
    const input = new appModules.Input(appModules.eventProcessor);
    const sessionManager = new appModules.SessionManager(
      appModules.eventProcessor,
      2
    );
    appModules.eventProcessor.emit('APP_STARTED');
    simulateTyping('sou');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    simulateTyping('falo');
    process.stdin.emit('keypress', {}, { name: 'return' });
    process.stdin.emit('keypress', {}, { name: 'n' });

    await sleep(50);

    const eventHistory = appModules.eventProcessor.eventHistory.map(
      (event: { event: any }) => event.event
    );
    expect(eventHistory).toMatchSnapshot();
    expect(output).toMatchSnapshot();
  });

});
