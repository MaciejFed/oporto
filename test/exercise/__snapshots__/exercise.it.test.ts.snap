// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercises IT Snapshot Fit In 1`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises IT Snapshot Fit In 2`] = `
Array [
  "Complete the sentence.",
  "Ela é  Angola.",
  "Ela é d Angola.",
  "Ela é de Angola.",
  "Correct! [voice]",
  "Ela é ",
  "d",
  "e",
  "Press key to conitinue...",
  "r - repeat the answer",
  "e - print explanation",
  "Complete the sentence.",
  "Eu  um chá.",
  "Eu w um chá.",
  "Eu wr um chá.",
  "Eu wro um chá.",
  "Eu wron um chá.",
  "Eu wrong um chá.",
  "Eu wronga um chá.",
  "Eu wrongan um chá.",
  "Eu wrongans um chá.",
  "Eu wrongansw um chá.",
  "Eu wronganswe um chá.",
  "Eu wronganswer um chá.",
  "Wrong! [voice]",
  "Eu ",
  "t",
  "o",
  "m",
  "o",
  "Press key to conitinue...",
  "r - repeat the answer",
  "e - print explanation",
]
`;

exports[`Exercises IT Snapshot NounTranslation 1`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises IT Snapshot NounTranslation 2`] = `
Array [
  "English: tea",
  "Portuguese: ",
  "Portuguese: o",
  "Portuguese: o ",
  "Portuguese: o c",
  "Portuguese: o ch",
  "Portuguese: o chá",
  "Correct! [voice]",
  "Portuguese: ",
  "o",
  " ",
  "c",
  "h",
  "á",
  "Press key to conitinue...",
  "r - repeat the answer",
  "Portuguese: o chá",
  "English: ",
  "English: t",
  "English: te",
  "English: tea",
  "Correct! [voice]",
  "English: ",
  "t",
  "e",
  "a",
  "Press key to conitinue...",
  "r - repeat the answer",
]
`;

exports[`Exercises IT Snapshot SentenceTranslation 1`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises IT Snapshot SentenceTranslation 2`] = `
Array [
  "English: How are you?",
  "Portuguese: ",
  "Portuguese: C",
  "Portuguese: Co",
  "Portuguese: Com",
  "Portuguese: Como",
  "Portuguese: Como ",
  "Portuguese: Como e",
  "Portuguese: Como es",
  "Portuguese: Como est",
  "Portuguese: Como está",
  "Portuguese: Como estás",
  "Portuguese: Como estás?",
  "Correct! [voice]",
  "Portuguese: ",
  "C",
  "o",
  "m",
  "o",
  " ",
  "e",
  "s",
  "t",
  "á",
  "s",
  "?",
  "Press key to conitinue...",
  "r - repeat the answer",
  "Portuguese: Como estás?",
  "English: ",
  "English: H",
  "English: Ho",
  "English: How",
  "English: How ",
  "English: How a",
  "English: How ar",
  "English: How are",
  "English: How are ",
  "English: How are y",
  "English: How are yo",
  "English: How are you",
  "English: How are you?",
  "Correct! [voice]",
  "English: ",
  "H",
  "o",
  "w",
  " ",
  "a",
  "r",
  "e",
  " ",
  "y",
  "o",
  "u",
  "?",
  "Press key to conitinue...",
  "r - repeat the answer",
]
`;

exports[`Exercises IT Snapshot Verb 1`] = `
Array [
  "APP_STARTED",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "EXERCISE_STARTED",
  "EXERCISE_DESCRIPTION_PRINTED",
  "EXERCISE_BODY_PRINTED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "KEY_PRESSSED",
  "ANSWER_SUBMITED",
  "ANSWER_CHECKED",
  "KEY_PRESSSED",
  "EXERCISE_NEXT",
  "APP_FINISHED",
]
`;

exports[`Exercises IT Snapshot Verb 2`] = `
Array [
  "Infinitive: ser",
  "Eu: ",
  "Eu: s",
  "Eu: so",
  "Eu: sou",
  "Correct! [voice]",
  "Eu: ",
  "s",
  "o",
  "u",
  "Press key to conitinue...",
  "r - repeat the answer",
  "Infinitive: falar",
  "Eu: ",
  "Eu: f",
  "Eu: fa",
  "Eu: fal",
  "Eu: falo",
  "Correct! [voice]",
  "Eu: ",
  "f",
  "a",
  "l",
  "o",
  "Press key to conitinue...",
  "r - repeat the answer",
]
`;
